addons/portal/static/src/js/portal_composer.js

```diff
@@ -145,8 +145,37 @@ var PortalComposer = publicWidget.Widget.extend({
      * @private
      * @returns {Promise}
      */
-    _onSubmitButtonClick: function () {
-        return new Promise(function (resolve, reject) {});
+    _onSubmitButtonClick: function (ev) {
+        var self = this,
+            $form = this.$el.find('form.o_portal_chatter_composer_form'),
+            route = $form.attr('action');
+        return new Promise(function (resolve, reject) {
+            if (route === '/mail/chatter_post') {
+                ev.preventDefault();
+                var data = $form.serializeArray();
+                self._rpc({
+                    route: route,
+                    params: _.object(_.pluck(data, 'name'), _.pluck(data, 'value')),
+                }).then(function (url) {
+                    var $parent = self.getParent();
+                    if (self.options.is_portal_chatter) {
+                        $parent._chatterInit().then(function (result) {
+                            // reset chatter widget
+                            if (self.options['display_composer']) {
+                                $parent._composer.destroy();
+                                $parent.start();
+                                $parent.renderElement();
+                                $parent._composer.replace($parent.$('.o_portal_chatter_composer'));
+                            }
+                        });
+                    }
+                    // for if chatter/composer open inside the BS modal
+                    if ($parent.$el.is('.modal.fade.modal_shown')) {
+                        window.location.reload();
+                    }
+                });
+            }
+        });
```

Comment:
Is there any reason to keep the composer as a form? You're getting the route from the form `action`, but since the template is  fixed, the route is always `'/mail/chatter_post'`, and then you're checking that it's the expected value when it can't be anything else.

Currently the form contains a lot of hidden inputs, so that they could be submitted by the form submit button, but since now we don't use the browser submit behaviour anymore, we can just keep these values in a JS object and clean up the template (most of the hidden inputs are available in the `options` field, and we have 2 separate hidden inputs for attachments, namely for id and access tokens, when we already have that information available in the attachments field)

The file input even had to be separated out of the original form since we upload the attachments immediately rather than on form submit.

This change would be a good opportunity to refactor the widget and template and clean things up, it would remove a lot of code and be less confusing than having a form that isn't really used as a form.

Reply:
===>  Is there any reason to keep the composer as a form? You're getting the route from the form action, but since the template is fixed, the route is always '/mail/chatter_post', and then you're checking that it's the expected value when it can't be anything else.


hello @sdegueldre , portal composer is extended in website-slides for rating on courses. If user reviews first time action is set to 'mail/chatter_post' but In case of modifying review the action url is  'slides/mail/update_comment' (https://github.com/odoo/odoo/blob/master/addons/website_slides/views/website_slides_templates_course.xml#L152) whcih is kept as 'POST' method. Hence we need to check url on 'submit button clicked' otherwise it will throw request type error

Reply:
@ktr-odoo that would need to be adapted as well, but this would also make the code cleaner:
* No longer need to replace the action attribute with a conditional in `portal_rating`'s `portal_chatter.xml`
* Instead, set the route as an option in `portal`'s `portal_composer.js` that defaults to `/mail/chatter_post`
* In `portal_rating`'s `rating_templates.xml` and in `website_slides`' `website_slides_template_courses.xml`, replace `force_submit_url` with `route`, which will then override the default route when the data-attributes are loaded into the options.

This will also allow you to stop reloading the page when a review is submitted or modified on a course in `website_slides` (but you will need to adapt that route too, and might need slightly different logic to reload the current ratings on the page than when reloading the chatter). This would make the use of the portal composer consistent across the board.

I understand that those changes are substantial, but I firmly believe that it will lead to code that's more maintainable and less confusing. I suggest that you do this on a copy of this branch, if you run into any other difficulties while making the changes we can discuss them then. I'm not completely opposed to the way you're currently doing things if there are too many obstacles, but I'd rather get rid of form if possible.